### PR TITLE
[CLEANUP beta] Abstract chainWatchers into an object.

### PR DESCRIPTION
**Pull request by krisselden**
Abstracting chainWatchers into an Object, cleans up some shape deopts, helps make stuff more clear.  This is good to go, I haven't squashed yet just so it is easier for people to review updates.

### CODE REVIEW

**Comment by mixonic** [comment]
@krisselden is green!

File: packages/ember-metal/lib/chains.js [hunk 1]
```diff
@@ -21,6 +21,90 @@ function Chains() { }
...
+        if (nodes[i] === node) {
+          return true;
+        }
+      }
+    }
+    return false;
+  },
+
+  revalidateAll() {
+    for (let key in this.nodes) {
```
Comment (stefanpenner) [inline packages/ember-metal/lib/chains.js]: do we ever delete from this.nodes?

Comment (krisselden) [inline packages/ember-metal/lib/chains.js]: no, we just empty the array

Comment (stefanpenner) [inline packages/ember-metal/lib/chains.js]: sounds good.

Comment (krisselden) [inline packages/ember-metal/lib/chains.js]: @stefanpenner should it just search a single array or linked list of well shaped entries?  like {key: 'foo', node: node} ?

Comment (stefanpenner) [inline packages/ember-metal/lib/chains.js]: array should be good.

**Comment by stefanpenner** [comment]
@krisselden whats blocking this/

File: packages/ember-metal/lib/chains.js [hunk 2]
```diff
@@ -21,6 +21,90 @@ function Chains() { }
 
 Chains.prototype = Object.create(null);
 
+// a map of nodes that reference a key
+function ChainWatchers(src) {
+  this.src = src;
+  this.nodes = Object.create(null);
```
Comment (mmun) [inline packages/ember-metal/lib/chains.js]:
```
function Empty() {}
Empty.prototype = Object.create(null);
...
this.nodes = new Empty();
```

**Comment by krisselden** [comment]
@stefanpenner rebased and squashed

**Pull request merged**